### PR TITLE
Introduce signature of ens name and gas Price when create contract

### PR DESCRIPTION
**Pull request by JustynaBroniszewska**

### CODE REVIEW

File: universal-login-commons/lib/core/utils/calculateSignature.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import {parseDomain} from '../utils/ens';
+
+export type InitializeWithENSArgs = {
+  ensName: string;
+  hashLabel: string;
+  node: string;
+  gasPrice: string;
+};
+
+export const calculateInitializeWithENSSignature = (privateKey: string, args: InitializeWithENSArgs) => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/calculateSignature.ts]: I would reverse order of arguments -> first what sign, than with what

File: universal-login-commons/lib/core/utils/calculateSignature.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import {Wallet, utils} from 'ethers';
+import {parseDomain} from '../utils/ens';
+
+export type InitializeWithENSArgs = {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/calculateSignature.ts]: Perhaps extract to seperate file `core/models/InitializeWithENSArgs`?

File: universal-login-commons/lib/core/utils/calculateSignature.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+export const calculateInitializeWithENSHash = (args: InitializeWithENSArgs) => utils.solidityKeccak256(
+  ['bytes32', 'string', 'bytes32', 'uint'],
+  [args.hashLabel, args.ensName, args.node, args.gasPrice]);
+
+export const calculateDeploySignature = (privateKey: string, ensName: string, gasPrice: string) => {
+  const [name] = parseDomain(ensName);
+  const hashLabel = utils.keccak256(utils.toUtf8Bytes(name));
+  const node = utils.namehash(ensName);
+  return calculateInitializeWithENSSignature(privateKey, {ensName, node, hashLabel, gasPrice});
+};
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/calculateSignature.ts]: 👍 Very nice helper file

File: universal-login-contracts/contracts/WalletMasterWithRefund.sol [hunk 4]
```diff
@@ -19,9 +19,42 @@ contract WalletMasterWithRefund is WalletMaster {
...
+        uint gasPrice,
+        bytes calldata signature) external onlyInitializing()
+    {
+        require(signature.length == 65, "Invalid signature");
+        require(isValidInitializeSignature(_key, _hashLabel, _name, _node, gasPrice, signature), "Invalid signature");
         this.initializeWithENS(_key, _hashLabel, _name, _node, ens, registrar, resolver);
         refund(getDeploymentGasUsed(), gasPrice, address(0), relayer);
     }
+
+    function isValidInitializeSignature(
```
Comment (marekkirejczyk) [inline universal-login-contracts/contracts/WalletMasterWithRefund.sol]: I would inline this function.

File: universal-login-commons/lib/core/models/InitializeWithENSArgs.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export type InitializeWithENSArgs = {
+  ensName: string;
+  hashLabel: string;
+  node: string;
+  gasPrice: string;
+};
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/models/InitializeWithENSArgs.ts]: 👍

File: universal-login-contracts/test/utils/utils.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import {utils} from 'ethers';
+
+export const switchENSNameInInitializeArgs = (initializeArgs: string[], name: string, domain = 'mylogin.eth') => {
+  const ensName = `${name}.${domain}`;
```
Comment (marekkirejczyk) [inline universal-login-contracts/test/utils/utils.ts]: `name` -> `label`?

File: universal-login-contracts/test/utils/utils.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  const hashLabel = utils.keccak256(utils.toUtf8Bytes(name));
+  const node = utils.namehash(ensName);
+  initializeArgs[1] = hashLabel;
+  initializeArgs[2] = ensName;
+  initializeArgs[3] = node;
+  return initializeArgs;
```
Comment (marekkirejczyk) [inline universal-login-contracts/test/utils/utils.ts]: `return [, hashLabel, ensName, node];`

File: universal-login-ops/src/dev/deployWalletMaster.js [hunk 8]
```diff
@@ -1,9 +1,10 @@
+import {utils} from 'ethers';
 import WalletMasterWithRefund from '@universal-login/contracts/build/WalletMasterWithRefund.json';
 import {getContractHash} from '@universal-login/commons';
 import {deployContract} from 'ethereum-waffle';
 
 export default async function deployWalletMaster(deployWallet) {
-  const {address} = await deployContract(deployWallet, WalletMasterWithRefund);
+  const {address} = await deployContract(deployWallet, WalletMasterWithRefund, [], {gasLimit: utils.bigNumberify('5000000')});
```
Comment (marekkirejczyk) [inline universal-login-ops/src/dev/deployWalletMaster.js]: Use defaultGas?

File: universal-login-ops/src/ops/deployMaster.ts [hunk 9]
```diff
@@ -6,14 +6,14 @@ import {deployContractAndWait} from '@universal-login/commons';
 
 export default async function deployMasterContract(wallet: Wallet) {
   console.log('Deploying wallet master contract...');
-  const overrides = {gasLimit: utils.bigNumberify(4000000)};
+  const overrides = {gasLimit: utils.bigNumberify(5000000)};
```
Comment (marekkirejczyk) [inline universal-login-ops/src/ops/deployMaster.ts]: Use defaultGas?

File: universal-login-ops/src/ops/deployMaster.ts [hunk 10]
```diff
@@ -6,14 +6,14 @@ import {deployContractAndWait} from '@universal-login/commons';
...
   console.log(`Wallet master contract address: ${contractAddress}`);
 }
 
 export async function deployMasterContractWithRefund(wallet: Wallet) {
   console.log(`Deploying wallet master with refund contract...`);
-  const overrides = {gasLimit: utils.bigNumberify(4000000)};
+  const overrides = {gasLimit: utils.bigNumberify(5000000)};
   const contractAddress = await deployContractAndWait(wallet, WalletMasterWithRefund, [], overrides);
   console.log(`Wallet master with refund contract address: ${contractAddress}`);
 }
```
Comment (marekkirejczyk) [inline universal-login-ops/src/ops/deployMaster.ts]: Use defaultGas?

File: universal-login-relayer/test/helpers/createWalletContract.js [hunk 11]
```diff
@@ -1,12 +1,12 @@
...
 import WalletProxy from '@universal-login/contracts/build/Proxy.json';
 import WalletMaster from '@universal-login/contracts/build/WalletMaster.json';
-import {ContractFactory, Contract} from 'ethers';
+import {ContractFactory, Contract, utils} from 'ethers';
 import {deployContract} from 'ethereum-waffle';
 import {encodeInitializeWithENSData} from '@universal-login/contracts';
 
 export default async function createWalletContract(wallet, ensService) {
-  const walletMaster = await deployContract(wallet, WalletMaster);
+  const walletMaster = await deployContract(wallet, WalletMaster, [], {gasLimit: utils.bigNumberify('5000000')});
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/createWalletContract.js]: Maybe extract `defaultOverridesForTests()`;

Comment (JustynaBroniszewska) [reply universal-login-relayer/test/helpers/createWalletContract.js]: I think that we could add `deployMaster` function (something like `deployFactory`)

**Pull request merged**